pageserver/src/tenant/timeline/detach_ancestor.rs

```diff
@@ -116,6 +125,404 @@ impl Default for Options {
     }
 }
 
+/// SharedState manages the pausing of background tasks (GC) for the duration of timeline detach
+/// ancestor.
+///
+/// Currently this is tracked at tenant level, but it could be moved to be on the roots
+/// of each timeline tree.
+#[derive(Default)]
+pub(crate) struct SharedState {
+    inner: std::sync::Mutex<SharedStateInner>,
+    gc_waiting: tokio::sync::Notify,
+    attempt_waiting: tokio::sync::Notify,
+}
+
+impl SharedState {
+    /// Notify an uninitialized shared state that an attempt to detach timeline ancestor continues
+    /// from previous instance.
+    pub(crate) fn notify(&self, attempt: &Attempt) {
+        self.inner.lock().unwrap().notify(attempt);
+    }
+
+    /// Only GC must be paused while a detach ancestor is ongoing. Compaction can happen, to aid
+    /// with any ongoing ingestion. Compaction even after restart is ok because layers will not be
+    /// removed until the detach has been persistently completed.
+    pub(crate) fn attempt_blocks_gc(&self) -> bool {
+        // if we have any started and not finished ancestor detaches, we must remain paused
+        // and also let any trying to start operation know that we've paused.
+
+        // Two cases:
+        // - there is an actual attempt started <-- handled
+        // - we have learned from indexparts that an attempt will be retried in near future
```

Comment:
second point is also handled via the `(ReadFromIndexPart, witnessed)` so an early gc (or one waiting for it background task can start should be able to mark it.